In [1]:
import tensorflow as tf
import pandas as pd

# 보스턴 집값 예측

## 1. 데이터 준비


In [2]:
파일경로 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/boston.csv'
보스턴 = pd.read_csv(파일경로)

## 종속변수, 독립변수 분리

In [3]:
독립 = 보스턴[['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax', 'ptratio', 'b', 'lstat']]
종속 = 보스턴[['medv']]
print(독립.shape, 종속.shape)

(506, 13) (506, 1)


## 2. 모델 생성

In [4]:
X = tf.keras.layers.Input(shape=[13])
# X를 입력층으로 받아 10개의 노드를 가진 히든레이어 생성
H = tf.keras.layers.Dense(10, activation='swish')(X)  # 히든레이어의 활성화 함수 swish
Y = tf.keras.layers.Dense(1)(H)  # 마지막 히든레이어에서 만들어진 H를 삽입해야 함
model = tf.keras.models.Model(X, Y)
model.compile(loss='mse')

#### 히든레이어
- 입력층과 출력층 사이의 연결고리인 히든레이어 
- 즉, 입력층과 출력층 사이에 하나의 층을 더 만듬
   - 입력층(506, 13), 히든레이어(506, 5), 출력층(506, 1)
   - 입력층에서 13개의 입력을 받고 히든레이어에서 5개의 출력을 받아 출력층에서 하나의 출력을 만들어 냄


## 모델 구조 확인


In [5]:
model.summary()
# 입력층 13개
# 10개의 노드 추가
# 출력층 1개

# param은 가중치로 입력층 13개와 bias 1개이므로 수식마다 14개의 항이 필요하고 항이 10개이므로 140
# 출력층은 10개의 노드와 1개의 bias로 인해 총 11

Model: "functional_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_1 (InputLayer)         [(None, 13)]              0         
_________________________________________________________________
dense (Dense)                (None, 10)                140       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 11        
Total params: 151
Trainable params: 151
Non-trainable params: 0
_________________________________________________________________


## 3. 모델 학습

In [6]:
model.fit(독립, 종속, epochs=100)

Epoch 1/100
16/16 [==============================] - 0s 3ms/step - loss: 1553.3019
Epoch 2/100
16/16 [==============================] - 0s 2ms/step - loss: 497.7584
Epoch 3/100
16/16 [==============================] - 0s 3ms/step - loss: 368.2353
Epoch 4/100
16/16 [==============================] - 0s 2ms/step - loss: 285.0297
Epoch 5/100
16/16 [==============================] - 0s 2ms/step - loss: 236.5855
Epoch 6/100
16/16 [==============================] - 0s 2ms/step - loss: 204.5943
Epoch 7/100
16/16 [==============================] - 0s 3ms/step - loss: 177.6093
Epoch 8/100
16/16 [==============================] - 0s 2ms/step - loss: 154.5627
Epoch 9/100
16/16 [==============================] - 0s 2ms/step - loss: 138.4224
Epoch 10/100
16/16 [==============================] - 0s 3ms/step - loss: 125.5881
Epoch 11/100
16/16 [==============================] - 0s 3ms/step - loss: 111.4421
Epoch 12/100
16/16 [==============================] - 0s 3ms/step - loss: 100.9716
Epoch 13/100

## 4. 모델 예측

In [7]:
print(model.predict(독립[:5]))
print(종속[:5])
# 24를 32로 예측, 21을 27로 예측

[[32.122787]
 [27.842491]
 [30.981773]
 [30.7305  ]
 [29.931023]]
   medv
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2


# 아이리스 품종 분류


## 1. 데이터 준비

In [8]:
파일경로 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv'
아이리스 = pd.read_csv(파일경로)

## 원핫인코딩

In [9]:
아이리스 = pd.get_dummies(아이리스)

## 종속변수, 독립변수 분리

In [10]:
독립 = 아이리스[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']]
종속 = 아이리스[['품종_setosa', '품종_versicolor', '품종_virginica']]
print(독립.shape, 종속.shape)

(150, 4) (150, 3)


In [11]:
X = tf.keras.layers.Input(shape=[4])
H = tf.keras.layers.Dense(8, activation="swish")(X) 
H = tf.keras.layers.Dense(8, activation="swish")(H)  
H = tf.keras.layers.Dense(8, activation="swish")(H)
Y = tf.keras.layers.Dense(3, activation='softmax')(H)
model = tf.keras.models.Model(X, Y)
model.compile(loss='categorical_crossentropy',
              metrics='accuracy')

## 모델 구조 확인

In [12]:
model.summary()

Model: "functional_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_2 (InputLayer)         [(None, 4)]               0         
_________________________________________________________________
dense_2 (Dense)              (None, 8)                 40        
_________________________________________________________________
dense_3 (Dense)              (None, 8)                 72        
_________________________________________________________________
dense_4 (Dense)              (None, 8)                 72        
_________________________________________________________________
dense_5 (Dense)              (None, 3)                 27        
Total params: 211
Trainable params: 211
Non-trainable params: 0
_________________________________________________________________


## 3. 모델 학습

In [13]:
model.fit(독립, 종속, epochs=100)

Epoch 1/100
5/5 [==============================] - 0s 4ms/step - loss: 1.2835 - accuracy: 0.2400
Epoch 2/100
5/5 [==============================] - 0s 3ms/step - loss: 1.1193 - accuracy: 0.3333
Epoch 3/100
5/5 [==============================] - 0s 3ms/step - loss: 1.0321 - accuracy: 0.3667
Epoch 4/100
5/5 [==============================] - 0s 3ms/step - loss: 0.9704 - accuracy: 0.4333
Epoch 5/100
5/5 [==============================] - 0s 4ms/step - loss: 0.9221 - accuracy: 0.5800
Epoch 6/100
5/5 [==============================] - 0s 4ms/step - loss: 0.8830 - accuracy: 0.6333
Epoch 7/100
5/5 [==============================] - 0s 3ms/step - loss: 0.8496 - accuracy: 0.6467
Epoch 8/100
5/5 [==============================] - 0s 3ms/step - loss: 0.8205 - accuracy: 0.6600
Epoch 9/100
5/5 [==============================] - 0s 4ms/step - loss: 0.7939 - accuracy: 0.6600
Epoch 10/100
5/5 [==============================] - 0s 3ms/step - loss: 0.7683 - accuracy: 0.6600
Epoch 11/100
5/5 [===========

## 4. 모델 예측

In [14]:
print(model.predict(독립[0:5]))
print(종속[0:5])

[[9.9967253e-01 3.2366751e-04 3.8430198e-06]
 [9.9874818e-01 1.2355706e-03 1.6228305e-05]
 [9.9938083e-01 6.0995301e-04 9.1366774e-06]
 [9.9871361e-01 1.2698447e-03 1.6558146e-05]
 [9.9973792e-01 2.5900529e-04 3.1460529e-06]]
   품종_setosa  품종_versicolor  품종_virginica
0          1              0             0
1          1              0             0
2          1              0             0
3          1              0             0
4          1              0             0
